frameworks.qmd

```diff
@@ -1,119 +1,1785 @@
+|-------------------------------|------------------------------------------------------------------------------------------------|-----------------------------------------------------------------|
+| Static (Declare-then-execute) | - Enable graph optimizations by seeing full model ahead of time<br>- Can export and deploy frozen graphs<br
+
+
+### Data Pipeline Tools
+
+Computational graphs can only be as good as the data they learn from and
+work on. Therefore, feeding training data efficiently is crucial for
+optimizing deep neural networks performance, though it is often
+overlooked as one of the core functionalities. Many modern AI frameworks
+provide specialized pipelines to ingest, process, and augment datasets
+for model training.
+
+Data Loaders
+
+At the core of these pipelines are data loaders, which handle reading
+examples from storage formats like CSV files or image folders. Reading
+training examples from sources like files, databases, object storage,
+etc. is the job of the data loaders. Deep learning models require
+diverse data formats depending on the application. Among the popular
+formats are CSV: A versatile, simple format often used for tabular data.
+TFRecord: TensorFlow\'s proprietary format, optimized for performance.
+Parquet: Columnar storage, offering efficient data compression and
+retrieval. JPEG/PNG: Commonly used for image data. WAV/MP3: Prevalent
+formats for audio data. For instance, tf.data is TensorFlows's
+dataloading pipeline: https://www.tensorflow.org/guide/data
+
+Data loaders batch examples to leverage vectorization support in
+hardware. Batching refers to grouping multiple data points for
+simultaneous processing, leveraging the vectorized computation
+capabilities of hardware like GPUs. While typical batch sizes range from
+32-512 examples, the optimal size often depends on the memory footprint
+of the data and the specific hardware constraints. Advanced loaders can
+stream virtually unlimited datasets from disk and cloud storage.
+Streaming large datasets from disk or networks instead of loading fully
+into memory. This enables virtually unlimited dataset sizes.
+
+Data loaders can also shuffle data across epochs for randomization, and
+preprocess features in parallel with model training to expedite the
+training process. Randomly shuffling the order of examples between
+training epochs reduces bias and improves generalization.
+
+Data loaders also support caching and prefetching strategies to optimize
+data delivery for fast, smooth model training. Caching preprocessed
+batches in memory so they can be reused efficiently during multiple
+training steps. Caching these batches in memory eliminates redundant
+processing. Prefetching, on the other hand, involves preloading
+subsequent batches, ensuring that the model never idles waiting for
+data.
+
```

Comment:
This data loaders section is great, but I think it can be consolidated a bit as some parts are a bit repetitive and it was split up slightly un-intuitively. From what I can tell the main points are that dataloaders (1) read the data from disk or memory, (2) allow for parallel computation on GPU via batching, (3) allow for shuffling (and I think augmentation should be mentioned in this part even though it is the next section), and (4) cacheing for efficiency. I think that points 1 and 4 can be consolidated, as well as 2 and 3 to simply say that dataloaders handle the fetching and cacheing of the inputs to the model, and can be used to batch, shuffle, and augment data before sending it through the model.